Create new Rollbar wrappers

javitonino opened the pull request:
Do not merge before https://github.com/CartoDB/cartodb/pull/6779
Closes #6778

This is a work in progress, here to discuss over code. This creates a new wrapper function CartoDB.log that allows:
- Report exceptions and/or messages. When both are specified, the exception shows up in Rollbar as `<ExceptionClassName>: <message>`, and the detail view shows the stacktrace.
- Specify logging levels: "critical", "error", "warning", "info", "debug", "ignore"
- Automatically adds person and request data. Person data can be overridden by passing the `user` parameter (for retrocompatibility). The same could be done for `request` but it does not seem to be any place in the code where `request` is anything other than `current_request`

The old wrappers have been rewritten to call this new one, to have a single point of access to the logging API (except for all Rollbar.* calls in the code). They behave exactly as before except for:
- More context data is added when available (actually a consequence of new rollbar gem)
- If specifying both message and exception to `report_exception`. Before, only the message was reported and the exception only showed in extra data. After, it will show the exception class and message in the listing, and the stacktrace in details.

cc @juanignaciosl @gfiorav

juanignaciosl commented on config/initializers/error_notifier.rb:
```diff
@@ -15,44 +15,47 @@
 
+  def self.log(level, exception: nil, message: nil, user: nil, **additional_data)
+    if Rails.env.development? || Rails.env.test?
+      ::Logger.new(STDOUT).error "#{level}: #{message}\n" + exception.inspect + "\n" + additional_data.inspect + "\n"
```
Log `user` and `exception.backtrace` (which is not read by `.inspect`) as well, please.

juanignaciosl commented on config/initializers/error_notifier.rb:
```diff
@@ -15,44 +15,47 @@
+  def self.rollbar_scope(user)
+    scope = {}
+    if user.respond_to?(:id)
+      scope['person'] = { 'id': user.id, 'username': user.username, 'email': user.email }
```
If Rollbar allows random attributes here, log the following ones as well, since they're really useful, please: `database_name`, `database_host`, `database_schema`,  `account_type`, `database_timeout`, `user_timeout`, and `google_sign_in`.

juanignaciosl commented:
Looks great. I think that some places are passing `user` inside `extra_options` (or similar arguments) and that user won't be passed to Rollbar `person`, but I'm not 100% sure.

I've added a couple of comments as well.

Is there a visible difference within old and new Rollbar traces with this code? If it is, some links with old and new examples would be great.

gfiorav commented:
Hey, sorry for the delay, I just read this.

Super luxurious to have this and it looks great, thanks :)

javitonino commented:
> I think that some places are passing user inside extra_options (or similar arguments) and that user won't be passed to Rollbar person, but I'm not 100% sure.

It works with the new wrapper. The function calls with `**additional_data` convert the hash to keyword arguments, and the user is correctly detected and reported.

I have also added more info to console logging and added more info about the user object to rollbar reports. This works when the user is manually specified. If the user is not given, Rollbar automatically (and very magically) adds it to the report, but only the id, username and email fields.

I tried to imitate what Rollbar does so we could have the user object in all reports, but I don't think it's worth it. It adds too much complexity and metaprogramming and we can always pass the user object when we need it.

cc @juanignaciosl

juanignaciosl commented:
:+1: :sparkles:

ethervoid commented:
Another thing I'll add is to remove all the Rollbar calls in the code: #6611

javitonino commented:
Retest this, please

javitonino commented:
Make sure to retest this after deploying new rollbar gem to production.

javitonino commented:
Try to add server (rails) to this.

javitonino commented:
Retest this, please

javitonino commented:
Okay, I think this is ready to merge (pending tests after updating this branch to master). One last quick CR please @juanignaciosl

javitonino commented:
I added a stack trace to reports without one (only messages). I also added some debugging information in case sending the report fails (see https://rollbar.com/vizzuality/CartoDB/items/15914/).

juanignaciosl commented on config/initializers/error_notifier.rb:
```diff
@@ -15,44 +15,72 @@
+    log('error', exception: e, message: message, **additional_data)
+  end
+
+  # Deprecated, use `log`
```
What do you think about adding `log_error`, `log_debug` and `log_warning` methods?

javitonino replied:
I'm not opposed to that, but I would create a module `CartoDB::Log` and name the mehtods by the level, so you call them like `CartoDB::Log.error(...)`.

juanignaciosl replied:
Oh, I'd _really_ love that. Invoking log methods on `CartoDB` has always been confusing to me.

juanignaciosl commented on config/initializers/error_notifier.rb:
```diff
@@ -15,44 +15,72 @@
   end
 
-  def self.notify_debug(message, additional_data={})
+  def self.log(level, exception: nil, message: nil, user: nil, **additional_data)
```
I'd add a `rescue` for this method so calling it can't fail at all.

juanignaciosl commented on config/initializers/error_notifier.rb:
```diff
@@ -15,44 +15,72 @@
+  # Creates a Rollbar scope that replaces the auto-detected person with the user passed as parameter
+  def self.rollbar_scope(user)
+    scope = {}
+    scope['person'] = user.respond_to?(:id)
```
How come you want `scope['person']` to be `true`/`false`?

javitonino replied:
That's wrong. Probably an attempt to use the ternary operator that I forgot to delete. Thanks!

juanignaciosl commented on config/initializers/error_notifier.rb:
```diff
@@ -15,44 +15,72 @@
+  def self.rollbar_scope(user)
+    scope = {}
+    scope['person'] = user.respond_to?(:id)
+    if !user.nil? && user.respond_to?(:id)
```
You can call `respond_to?` on `nil`, so nil check is not needed.

juanignaciosl commented:
Ping me again after the changes. I'll really love the `CartoDB::Log` module ^_^

javitonino commented:
I've moved the new logging functions to `Carto::Logger` (Carto::Log is the import log model). The new calls look like: `Carto::Logger.error(exception: e, message: 'some message', extra: 'wadus')`

cc @juanignaciosl

juanignaciosl commented on lib/cartodb/logger.rb:
```diff
@@ -1,6 +1,71 @@
+  end
+
+  # Deprecated logging
+  class StdoutLogger
```
About this refactor: I've found one occurrence that might be related at `./spec/models/geocoding_spec.rb`: `Logger.any_instance.expects(:info).once {|str|`. Since it's an spec it might be relevant, but maybe it's related to `::Logger`, not the refactored one. Take a look if you haven't done it, please.

juanignaciosl commented on lib/cartodb/logger.rb:
```diff
@@ -1,6 +1,71 @@
+    def self.log(level, exception: nil, message: nil, user: nil, **additional_data)
+      # Include the call stack if not already present
+      if exception.nil? && additional_data[:stack].nil?
+        additional_data[:stack] = caller
```
<3

juanignaciosl commented on lib/cartodb/logger.rb:
```diff
@@ -1,6 +1,71 @@
+        Rollbar.log('warning', nil, 'Could not report to Rollbar', stack: caller)
+      end
+    rescue
+      # Do not bubble errors generated during exception reporting
```
I don't want to be picky, but maybe I'd change this to

```ruby
rescue => e
  begin Rollbar.error(e) rescue end
end
```

Since _nothing_ should really call this code, if we missed something we'll notice it, and if something's wrong with Rollbar it won't fail.

Up to you though.

juanignaciosl commented:
I've dropped a few comments. Nothing major, take into account what you consider and :+1:
